File: src/main/java/voidchess/helper/Move.java
package voidchess.helper;

/**
 * @author stephan
 */
public class Move {
    final private static Move[] moves = new Move[64*64];

    static {
        for (int moveIndex = 0; moveIndex < moves.length; moveIndex++) {
            int fromIndex = moveIndex % 64;
            int toIndex = moveIndex / 64;
            Position fromPos = Position.Companion.byIndex(fromIndex);
            Position toPos = Position.Companion.byIndex(toIndex);
            moves[moveIndex] = new Move(fromPos, toPos);
        }
    }

    final public Position from;
    final public Position to;

    public static Move get(Position from, Position to) {
        return moves[getMoveIndex(from.getIndex(), to.getIndex())];
    }

    public static Move byCode(String code) {
        int fromColumn = (int) code.charAt(0) - 97;
        int fromRow = (int) code.charAt(1) - 49;
        int toColumn = (int) code.charAt(3) - 97;
        int toRow = (int) code.charAt(4) - 49;

        return get(
                Position.Companion.get(fromRow, fromColumn),
                Position.Companion.get(toRow, toColumn)
        );
    }

    public static boolean isValid(String code) {
        if (code == null || code.length() != 5 || code.charAt(2) != '-') {
            return false;
        }

        int fromColumn = (int) code.charAt(0) - 97;
        int fromRow = (int) code.charAt(1) - 49;
        int toColumn = (int) code.charAt(3) - 97;
        int toRow = (int) code.charAt(4) - 49;

        return liesIn07Range(fromColumn) &&
                liesIn07Range(fromRow) &&
                liesIn07Range(toColumn) &&
                liesIn07Range(toRow);
    }

    private static boolean liesIn07Range(int number) {
        return number >= 0 && number <= 7;
    }

    // optimized from: fromIndex + toIndex * 64
    private static int getMoveIndex(int fromIndex, int toIndex) {
      return fromIndex + (toIndex<<6);
    }

    protected Move(Position from, Position to) {
        this.from = from;
        this.to = to;
    }

    public String toString() {
        StringBuffer buffer = new StringBuffer(5);
        buffer.append(from.toString());
        buffer.append("-");
        buffer.append(to.toString());
        return buffer.toString();
    }

    public boolean equalsMove(Move other) {
        return toString().equals(other.toString());
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Move)) return false;
        return equalsMove((Move) obj);
    }

    @Override
    public int hashCode() {
        return to.hashCode();
    }
}

File: src/main/kotlin/voidchess/helper/Move.kt
package voidchess.helper

/**
 * @author stephan
 */
class Move private constructor(val from: Position, val to: Position) {
    private val index = getMoveIndex(from.index, to.index)

    fun equalsMove(move: Move) = this === move //index == move.index
    override fun equals(other: Any?) = this === other //other is Move && index == other.index
    override fun toString() = "$from-$to"
    override fun hashCode() = index

    companion object {
        private val moves = Array(64 * 64) {
            val fromIndex = it % 64
            val toIndex = it / 64
            return@Array Move(
                    Position.byIndex(fromIndex),
                    Position.byIndex(toIndex)
            )
        }

        @JvmStatic
        operator fun get(from: Position, to: Position) = moves[getMoveIndex(from.index, to.index)]

        @JvmStatic
        fun byCode(code: String): Move {
            val fromColumn = code[0].toInt() - 97
            val fromRow = code[1].toInt() - 49
            val toColumn = code[3].toInt() - 97
            val toRow = code[4].toInt() - 49

            return get(
                    Position.get(fromRow, fromColumn),
                    Position.get(toRow, toColumn)
            )
        }

        @JvmStatic
        fun isValid(code: String?): Boolean {
            if (code == null || code.length != 5 || code[2] != '-') {
                return false
            }

            val fromColumn = code[0].toInt() - 97
            val fromRow = code[1].toInt() - 49
            val toColumn = code[3].toInt() - 97
            val toRow = code[4].toInt() - 49

            return fromColumn in 0..7 &&
                    fromRow in 0..7 &&
                    toColumn in 0..7 &&
                    toRow in 0..7
        }

        // optimized from: fromIndex + toIndex * 64
        private fun getMoveIndex(fromIndex: Int, toIndex: Int) = fromIndex + (toIndex shl 6)
    }
}
